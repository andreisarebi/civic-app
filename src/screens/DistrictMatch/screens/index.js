export { default as Address } from './Address';
export { default as ZipCode } from './ZipCode';
export { default as Success } from './Success';
export { default as Failure } from './Failure';
