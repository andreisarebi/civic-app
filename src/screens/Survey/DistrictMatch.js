import React from 'react';
import { View, Text, StyleSheet, Image, TextInput } from 'react-native';
import Mixins from '../../styles/mixins';
import colors from '../../styles/colors';

import ButtonSubmit from '../components/ButtonSubmit';

const zipCodeImage = require('../../assets/images/district-match-zip.png');

const DistrictMatch = () => {
  return(
    <View style={styles.container}>
      <Text style={styles.header}>Just one more thing</Text>
      <Image style={styles.image} source={zipCodeImage}/>
      <Text style={styles.bodyText}>We need your zip code to find your district.</Text>
      <View style={styles.inputContainer}>
        <TextInput
          autoCapitalize="none"
          autoCorrect={false}
          autoFocus={false}
          clearButtonMode="while-editing"
          keyboardType="numeric"
          style={styles.textInput}
          placeholder="ZIP code"
        />
        <ButtonSubmit/>
      </View>
    </View>
  )
}


        // onChangeText={email => this.setState({ email })}
        // value={this.state.email}

const styles = StyleSheet.create({
  container: {
    backgroundColor: 'white',
    borderRadius: 15,
    margin: 25,
    flexDirection: 'column',
    alignItems: 'center',
    width: '90%',
    ...Mixins.shadow
  },
  header: {
    fontSize: 28,
    paddingTop: 20
  },
  bodyText: {
    fontSize: 18,
    width: 230,
    textAlign: 'center'
  },
  image: {
    marginTop: 25,
    marginBottom:20
  },
  textInput: {
    color: colors.black,
    fontSize: 18,
    paddingTop: 10,
    paddingBottom: 10
  },
  inputContainer: {
    flexDirection: 'column',
    justifyContent: 'center',
    width: '90%',
    padding: 20,
  },
})

export default DistrictMatch;
