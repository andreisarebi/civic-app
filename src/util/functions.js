import distanceInWordsStrict from 'date-fns/distance_in_words_strict';

export const distanceInWordsFromNow = (date) => {
  return distanceInWordsStrict(date, Date.now());
}
