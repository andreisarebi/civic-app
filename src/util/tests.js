import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import { Provider } from 'react-redux';
import store from '../../App/configureStore';

export const createSnapshot = (Component, props) => {
  const renderer = new ShallowRenderer();
  renderer.render(<Component {...props} />);
  return renderer.getRenderOutput();
};

export const createConnectedSnapshot = (Component, props) => {
  const renderer = new ShallowRenderer();
  renderer.render(<Provider store={store}><Component {...props} /></Provider>);
  return renderer.getRenderOutput();
};
