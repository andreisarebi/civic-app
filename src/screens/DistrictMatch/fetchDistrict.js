import { buildCivicQuery } from '../../util/functions';

export default async function(){
  const url = buildCivicQuery([this.state.address, this.state.city, this.state.zipcode]);
  try {
    const result = await fetch(url);
    const resultJson = await result.json();
    if(!resultJson.error){
      if(!resultJson.offices){
        if(this.state.view==='ADDRESS_NEEDED'){
          this.setState({view: 'DISTRICT_NOT_FOUND'}, ()=>{
            this.setState({isLoading: false});
          })
        }else{
          this.setState({view: 'ADDRESS_NEEDED'}, () => {
            this.setState({isLoading: false});
          });
        }
      }else{
        const officeName = resultJson.offices[0].name.split(' ');
        const district = officeName[officeName.length-1];
        /**
         * Do something with the district here then navigate to main App...
         */
        this.props.navigate('App');
      }
    }else{
      throw resultJson.error;
    }
  } catch(error){
    if(this.state.view==='ADDRESS_NEEDED'){
      this.setState({view: 'DISTRICT_NOT_FOUND'}, ()=>{
        this.setState({isLoading: false});
      })
    }else{
      this.setState({view: 'ADDRESS_NEEDED'}, () => {
        this.setState({isLoading: false});
      });
    }
  }
}
