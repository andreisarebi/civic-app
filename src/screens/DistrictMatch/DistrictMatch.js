import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { View, StyleSheet, ActivityIndicator } from 'react-native';

// Views
import {ZipCode, Address, Success, Failure} from './screens';
// API Method & Controller
import fetchDistrict from './fetchDistrict';
// styles
import Mixins from '../../styles/mixins';
import colors from '../../styles/colors';


class DistrictMatch extends Component {
  constructor(props){
    super(props);
    // Bind external fetchDistrict async function
    // acts as controller for component views, fetching districts, and handling errors
    this.fetchDistrict = fetchDistrict.bind(this);
  }

  state = {
    // params
    zipcode: '',
    address: '',
    city: '',
    // loading state
    isLoading: false,
    // view to load via switch statement in renderScreen method
    // options are ENTER_ZIPCODE, ADDRESS_NEEDED, DISTRICT_FOUND, and DISTRICT_NOT_FOUND
    view:'ENTER_ZIPCODE',
    // district response parsed from Google Civic API to be sent to elections screen
    district: '',
    // Error message in case of badly parsed request to API
    error: ''
  };

  submit = () => {
    // sets loading state and resets error state
    this.setState({isLoading: true, error: ''}, () => {
      // calls fetchDistrict in callback function so loading state can be set first
      this.fetchDistrict();
    })
  }

  handleChange = (input) => {
    // accepts an argument of an object with the state property and it's value
    // e.g. {city: 'Baltimore'}
    this.setState({...input});
  }

  renderScreen = () => {
    // pass main component render methods down to screen views
    const methods = {
      handleChange: this.handleChange,
      submit: this.submit,
      navigate: this.props.navigate
    };
    // control views based on view state
    switch(this.state.view){
      case 'ENTER_ZIPCODE':
        return <ZipCode styles={styles} {...this.state} {...methods} />;
      case 'ADDRESS_NEEDED':
        return <Address styles={styles} {...this.state} {...methods} />;
      case 'DISTRICT_FOUND':
        return <Success district={this.state.district} {...methods} />;
      case 'DISTRICT_NOT_FOUND':
        return <Failure {...methods}/>;
      // this.state.view should not be null
      default:
        return null;
    }
  }

  render(){
    const { zipcode, address, city, isLoading, view, district, error } = this.state;
    if(isLoading){
      return (
        <View>
          <ActivityIndicator size="large" color="#0000ff" />
        </View>
      )
    }
    return (
      <View style={styles.container}>
        {this.renderScreen()}
      </View>
    )
  }
}

// navigate function passed from index.js
DistrictMatch.propTypes = {
  navigate: PropTypes.func
}

export const styles = StyleSheet.create({
  container: {
    backgroundColor: 'white',
    borderRadius: 15,
    margin: 25,
    flexDirection: 'column',
    alignItems: 'center',
    width: '90%',
    ...Mixins.shadow
  },
  header: {
    fontSize: 28,
    paddingTop: 20
  },
  bodyText: {
    fontSize: 18,
    width: 230,
    paddingTop: 20,
    textAlign: 'center'
  },
  image: {
    marginTop: 25,
    marginBottom:20
  },
  textInput: {
    color: colors.black,
    fontSize: 18,
    paddingTop: 10,
    paddingBottom: 10
  },
  inputContainer: {
    flexDirection: 'column',
    justifyContent: 'center',
    width: '90%',
    padding: 20,
  },
  errorText: {
    color: colors.orange
  }
})

export default DistrictMatch;
