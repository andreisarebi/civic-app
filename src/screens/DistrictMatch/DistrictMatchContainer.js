import { connect } from 'react-redux';
import DistrictMatch from './DistrictMatch';
import { saveDistrict } from '../../user/redux';

const mapDispatchToProps = dispatch => ({
  saveDistrict: district => dispatch(saveDistrict(district)),
});

export default connect(
  null,
  mapDispatchToProps,
)(DistrictMatch);
