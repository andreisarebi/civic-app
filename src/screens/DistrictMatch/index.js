import React, {Component} from 'react';
import { KeyboardAvoidingView, Button, StyleSheet } from 'react-native';
import DistrictMatch from './DistrictMatch';

class DistrictMatchScreen extends Component {
  render() {
    return (
      <KeyboardAvoidingView style={styles.keyboardAvoidView} behavior="position">
        <DistrictMatch navigate={this.props.navigation.navigate}/>
      </KeyboardAvoidingView>
    );
  }
}

const styles = StyleSheet.create({
  keyboardAvoidView: {
    flex: 1,
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'center',
    width: '100%'
  }
})

export default DistrictMatchScreen;
