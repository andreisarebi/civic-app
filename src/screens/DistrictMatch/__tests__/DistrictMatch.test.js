import React from 'react';
import { Provider } from 'react-redux';
import { createSnapshot } from '../../../util/tests';
import DistrictMatch from '../DistrictMatchContainer';
import store from '../../../../App/configureStore';

const DistrictMatchWithStore = () => (
  <Provider store={store}>
    <DistrictMatch/>
  </Provider>
);

describe('DistrictMatch', () => {
  it('renders correctly', () => {
    const snapshot = createSnapshot(DistrictMatchWithStore);
    expect(snapshot).toMatchSnapshot();
  });
});