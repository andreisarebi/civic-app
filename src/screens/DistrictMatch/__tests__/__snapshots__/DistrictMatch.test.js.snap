// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DistrictMatch renders correctly 1`] = `
<Component
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "white",
      "borderRadius": 15,
      "elevation": 1,
      "flexDirection": "column",
      "margin": 25,
      "minWidth": "90%",
      "shadowColor": "black",
      "shadowOffset": Object {
        "height": 2,
        "width": 0,
      },
      "shadowOpacity": 0.2,
      "shadowRadius": 3
    }
  }
>
  <ZipCode
    address=""
    city=""
    district=""
    error=""
    handleChange={[Function]}
    isLoading={false}
    navigate={undefined}
    styles={
      Object {
        "bodyText": Object {
          "fontSize": 18,
          "paddingTop": 20,
          "textAlign": "center",
          "width": 230,
        },
        "container": Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderRadius": 15,
          "elevation": 1,
          "flexDirection": "column",
          "margin": 25,
          "minWidth": "90%",
          "shadowColor": "black",
          "shadowOffset": Object {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.2,
          "shadowRadius": 3
        },
        "errorText": Object {
          "color": "#ff5722",
        },
        "header": Object {
          "fontSize": 28,
          "paddingTop": 20,
        },
        "image": Object {
          "marginBottom": 20,
          "marginTop": 25,
        },
        "inputContainer": Object {
          "flexDirection": "column",
          "justifyContent": "center",
          "padding": 20,
          "width": "90%",
        },
        "textInput": Object {
          "color": "#000000",
          "fontSize": 18,
          "paddingBottom": 10,
          "paddingTop": 10,
        },
      }
    }
    submit={[Function]}
    view="ENTER_ZIPCODE"
    zipcode=""
  />
</Component>
`;
